File: agentic-clj/test/agentic/windows_test.clj
(ns agentic.windows-test
  "Phase 3: portable keyed time-windows. Sliding (the VelocityDetector), tumbling, and session
   aggregates. Mirrors jagentic-core's org.jagentic.core.windows.* goldens — behaviour at parity."
  (:require [clojure.test :refer [deftest is testing]]
            [agentic.windows :as w]))

(deftest sliding-count
  (testing "count within (ts-window, ts]; independent per key; old events evicted"
    (let [sw (w/sliding-window 60000)]
      (doseq [ts [1000 2000 3000 4000]]
        (w/slide-add sw "h1" ts))
      (is (= 5 (:count (w/slide-add sw "h1" 5000))) "all 5 events inside the 60s window")
      (is (= 1 (:count (w/slide-add sw "h1" 200000))) "only the latest survives the eviction")
      (is (= 1 (:count (w/slide-add sw "h2" 1000))) "a fresh key is independent"))))

(deftest sliding-sum
  (testing "sum of values within the window"
    (let [sw (w/sliding-window 10000)]
      (w/slide-add sw "acct" 0 100.0)
      (let [agg (w/slide-add sw "acct" 5000 250.0)]
        (is (= 2 (:count agg)))
        (is (= 350.0 (:sum agg)))))))

(deftest tumbling-buckets
  (testing "events accumulate into floor(ts/window) buckets; later-bucket events emit the prior"
    (let [tw (w/tumbling-window 1000)]
      (is (nil? (w/tumble-add tw "k" 100)) "first event opens bucket 0, nothing closed")
      (is (nil? (w/tumble-add tw "k" 200)) "still bucket 0")
      (let [closed (w/tumble-add tw "k" 1500)]
        (is (= 0 (:start closed)) "bucket 0 starts at 0")
        (is (= 2 (:count closed)) "two events in bucket 0"))
      (let [closed (w/tumble-close tw "k")]
        (is (= 1000 (:start closed)) "open bucket 1 starts at 1000")
        (is (= 1 (:count closed)) "one event in bucket 1")))))

(deftest session-grouping
  (testing "events within gap join a session; a larger gap closes (emits) the prior session"
    (let [sw (w/session-window 5000)]
      (is (nil? (w/session-add sw "u" 0)) "first event opens a session")
      (is (nil? (w/session-add sw "u" 2000)) "2000-0 <= gap, same session")
      (let [closed (w/session-add sw "u" 10000)]
        (is (= 0 (:start closed)))
        (is (= 2000 (:end closed)))
        (is (= 2 (:count closed)) "the prior session held two events"))
      (is (= 1 (:count (w/session-close sw "u"))) "the new session holds the 10000 event")
      (is (nil? (w/session-close sw "u")) "nothing open after close"))))

File: ports/go/core/windows_test.go
package core

import "testing"

func TestSlidingWindowCount(t *testing.T) {
	w := NewSlidingWindow(60000)
	var last WindowState
	for ts := int64(1000); ts <= 5000; ts += 1000 {
		last = w.AddCount("h1", ts)
	}
	if last.Count != 5 {
		t.Fatalf("sliding count: want 5, got %d", last.Count)
	}
	// Far in the future — all prior entries evicted, only this one remains.
	if got := w.AddCount("h1", 200000); got.Count != 1 {
		t.Fatalf("sliding count after gap: want 1, got %d", got.Count)
	}
	// Independent per key.
	if got := w.AddCount("h2", 1000); got.Count != 1 {
		t.Fatalf("sliding count h2: want 1, got %d", got.Count)
	}
}

func TestSlidingWindowSum(t *testing.T) {
	w := NewSlidingWindow(10000)
	w.Add("acct", 0, 100)
	got := w.Add("acct", 5000, 250)
	if got.Count != 2 {
		t.Fatalf("sliding sum count: want 2, got %d", got.Count)
	}
	if got.Sum != 350 {
		t.Fatalf("sliding sum: want 350, got %v", got.Sum)
	}
}

func TestTumblingWindow(t *testing.T) {
	w := NewTumblingWindow(1000)
	if _, ok := w.AddCount("k", 100); ok {
		t.Fatalf("tumbling @100: want ok=false")
	}
	if _, ok := w.AddCount("k", 200); ok {
		t.Fatalf("tumbling @200: want ok=false")
	}
	b, ok := w.AddCount("k", 1500)
	if !ok {
		t.Fatalf("tumbling @1500: want ok=true")
	}
	if b.Start != 0 || b.Count != 2 {
		t.Fatalf("tumbling emitted bucket: want {Start:0, Count:2}, got %+v", b)
	}
	cb, ok := w.Close("k")
	if !ok {
		t.Fatalf("tumbling close: want ok=true")
	}
	if cb.Start != 1000 || cb.Count != 1 {
		t.Fatalf("tumbling close bucket: want {Start:1000, Count:1}, got %+v", cb)
	}
}

func TestSessionWindow(t *testing.T) {
	w := NewSessionWindow(5000)
	if _, ok := w.AddCount("u", 0); ok {
		t.Fatalf("session @0: want ok=false")
	}
	if _, ok := w.AddCount("u", 2000); ok {
		t.Fatalf("session @2000: want ok=false")
	}
	s, ok := w.AddCount("u", 10000)
	if !ok {
		t.Fatalf("session @10000: want ok=true")
	}
	if s.Start != 0 || s.End != 2000 || s.Count != 2 {
		t.Fatalf("session emitted: want {Start:0, End:2000, Count:2}, got %+v", s)
	}
	cs, ok := w.Close("u")
	if !ok {
		t.Fatalf("session close: want ok=true")
	}
	if cs.Count != 1 {
		t.Fatalf("session close count: want 1, got %d", cs.Count)
	}
	if _, ok := w.Close("u"); ok {
		t.Fatalf("session close again: want ok=false")
	}
}
